### ISSUE TITLE
[BUG] The cached_loss_mask cannot be consistent

### ISSUE THREAD

**Issue by XLzed** [NVIDIA/Megatron-LM#1298]
**How to make training resumption bitwise reproducible?**
When setting the environment variables **NVTE_ALLOW_NONDETERMINISTIC_ALGO=0, NCCL_ALGO=Ring**, training can be bitwise reproducible by loading same checkpoint. 
Training resume from an intermediate checkpoint, the loss curve can be nearly same as complete training from original checkpoint, but the precision of the loss can no longer be aligned in bitwise. What might be causing the problem and how to make training resumption bitwise reproducible?

**Comment by XLzed** [comment]
While masks_and_position_ids_are_cacheable is True, the cached_loss_mask should be the same value all the time. But the loss_mask and cached_loss_mask reference each other, which makes the loss_mask incorrect after first iteration, torch.sum(loss_mask) will be monotone decreasing. This is also why the loss curve of training resumption can not be bitwise-aligned with the complete training.
The bug seems introduced from core-0.7.0, which zero loss_mask by _pad_token_id.
**The original code with bug.**
```python
        if (
            not self.masks_and_position_ids_are_cacheable
            or not self.masks_and_position_ids_are_cached
        ):
            attention_mask, loss_mask, position_ids = _get_ltor_masks_and_position_ids(
                tokens,
                self.config.tokenizer.eod,
                self.config.reset_position_ids,
                self.config.reset_attention_mask,
                self.config.eod_mask_loss,
                self.config.create_attention_mask,
            )
            if self.masks_and_position_ids_are_cacheable:
                self.cached_attention_mask = attention_mask
                self.cached_loss_mask = loss_mask
                self.cached_position_ids = position_ids
                self.masks_and_position_ids_are_cached = True
        else:
            attention_mask = self.cached_attention_mask
            loss_mask = self.cached_loss_mask
            position_ids = self.cached_position_ids

        # For padded sequences, mask the loss
        loss_mask[labels == self._pad_token_id] = 0.0
```

**Bug fixed, use torch.clone() to avoid modification of cached_loss_mask.**
```python
        if (
            not self.masks_and_position_ids_are_cacheable
            or not self.masks_and_position_ids_are_cached
        ):
            attention_mask, loss_mask, position_ids = _get_ltor_masks_and_position_ids(
                tokens,
                self.config.tokenizer.eod,
                self.config.reset_position_ids,
                self.config.reset_attention_mask,
                self.config.eod_mask_loss,
                self.config.create_attention_mask,
            )
            if self.masks_and_position_ids_are_cacheable:
                self.cached_attention_mask = attention_mask
                self.cached_loss_mask = torch.clone(loss_mask)
                self.cached_position_ids = position_ids
                self.masks_and_position_ids_are_cached = True
        else:
            attention_mask = self.cached_attention_mask
            loss_mask = torch.clone(self.cached_loss_mask)
            position_ids = self.cached_position_ids

        # For padded sequences, mask the loss
        loss_mask[labels == self._pad_token_id] = 0.0
```

**Comment by XLzed** [comment]
My custom tokenizer is used incorrectly, the pad_id should not be set. Close this comment.

**Comment by rmwor1d** [comment]
@XLzed hi, same question here, but there are indeed tokenizers contain a pad_id. Is your solution (using default -1 instead of pad_id to pad sequences) supposed to be a temporary approach to bypass this problem? Thx
